File: src/main/java/kr/dogfoot/hwpxlib/object/metainf/ManifestXMLFile.java
package kr.dogfoot.hwpxlib.object.metainf;

import kr.dogfoot.hwpxlib.object.common.ObjectType;
import kr.dogfoot.hwpxlib.object.common.SwitchableObject;

import java.util.ArrayList;

public class ManifestXMLFile extends SwitchableObject {
    private final ArrayList<FileEntry> fileEntryList;

    public ManifestXMLFile() {
        fileEntryList = new ArrayList<FileEntry>();
    }

    @Override
    public ObjectType _objectType() {
        return ObjectType.odf_manifest;
    }

    public int countOfFileEntry() {
        return fileEntryList.size();
    }

    public FileEntry getFileEntry(int index) {
        return fileEntryList.get(index);
    }

    public int getFileEntryIndex(FileEntry fileEntry) {
        int count = fileEntryList.size();
        for (int index = 0; index < count; index++) {
            if (fileEntryList.get(index) == fileEntry) {
                return index;
            }
        }
        return -1;
    }

    public void addFileEntry(FileEntry fileEntry) {
        fileEntryList.add(fileEntry);
    }

    public FileEntry addNewFileEntry() {
        FileEntry fileEntry = new FileEntry();
        fileEntryList.add(fileEntry);
        return fileEntry;
    }

    public void insertFileEntry(FileEntry fileEntry, int position) {
        fileEntryList.add(position, fileEntry);
    }

    public void removeFileEntry(int position) {
        fileEntryList.remove(position);
    }

    public void removeFileEntry(FileEntry fileEntry) {
        fileEntryList.remove(fileEntry);
    }

    public Iterable<FileEntry> fileEntries() {
        return fileEntryList;
    }

    public ManifestXMLFile clone() {
        ManifestXMLFile cloned = new ManifestXMLFile();
        cloned.copyFrom(this);
        return cloned;
    }

    public void copyFrom(ManifestXMLFile from) {
        for (FileEntry fileEntry : from.fileEntryList) {
            fileEntryList.add(fileEntry.clone());
        }
        super.copyFrom(from);
    }
}

File: js/src/object/metainf/ManifestXMLFile.js
import { SwitchableObject } from '../common/SwitchableObject.js';
import { ObjectType } from '../common/ObjectType.js';

/**
 * File entry in manifest
 */
export class FileEntry {
  constructor() {
    this._fullPath = null;
    this._mediaType = null;
    this._size = null;
  }

  _objectType() {
    return ObjectType.odf_file_entry;
  }

  fullPath() { return this._fullPath; }
  fullPathSet(value) { this._fullPath = value; }
  fullPathAnd(value) { this._fullPath = value; return this; }

  mediaType() { return this._mediaType; }
  mediaTypeSet(value) { this._mediaType = value; }
  mediaTypeAnd(value) { this._mediaType = value; return this; }

  size() { return this._size; }
  sizeSet(value) { this._size = value; }
  sizeAnd(value) { this._size = value; return this; }

  clone() {
    const cloned = new FileEntry();
    cloned._fullPath = this._fullPath;
    cloned._mediaType = this._mediaType;
    cloned._size = this._size;
    return cloned;
  }
}

/**
 * META-INF/manifest.xml file representation
 */
export class ManifestXMLFile extends SwitchableObject {
  constructor() {
    super();
    this._fileEntries = [];
  }

  _objectType() {
    return ObjectType.odf_manifest;
  }

  fileEntries() {
    return this._fileEntries;
  }

  addFileEntry() {
    const entry = new FileEntry();
    this._fileEntries.push(entry);
    return entry;
  }

  clone() {
    const cloned = new ManifestXMLFile();
    cloned.copyFrom(this);
    return cloned;
  }

  copyFrom(from) {
    this._fileEntries = [];
    for (const entry of from._fileEntries) {
      this._fileEntries.push(entry.clone());
    }
    this._copyFromSwitchable(from);
  }
}

export default ManifestXMLFile;
